design/appfile-design.md

```diff
@@ -0,0 +1,337 @@
+# Appfile: Extensible, User-friendly Application Config Format
+
+- Owner: Hongchao Deng (@hongchaodeng)
+- Date: 10/14/2020
+- Status: Implemented
+
+## Table of Contents
+
+- [Intro](#intro)
+- [Goals](#goals)
+- [Proposal](#proposal)
+  - [Registration via Definition/Capability](#registration-via-definitioncapability)
+  - [Templating](#templating)
+  - [CLI/UI interoperability](#cliui-interoperability)
+  - [vela up](#vela-up)
+- [Examples](#examples)
+
+## Intro
+
+Vela supports a user-friendly `docker-compose` style config format called `Appfile`. It allows you to define an application's workloads and traits with an opinionated, simplified API interface.
+
+Here's an example to deploy a NodeJS express service:
+
+```yaml
+services:
+  express-server:
+    build:
+      image: oamdev/testapp:v1
+      docker:
+        file: Dockerfile
+        context: .
+
+    cmd: ["node", "server.js"]
+
+    route:
+      domain: example.com
+      http: # match the longest prefix
+        "/": 8080
+
+    env:
+      - FOO=bar
+      - FOO2=sec:my-secret # map the key same as the env name (`FOO2`) from my-secret to env var
+      - FOO3=sec:my-secret:key # map specific key from my-secret to env var
+      - sec:my-secret # map all KV pairs from my-secret to env var
+
+    files: # Mount secret as a file
+      - /mnt/path=sec:my-secret
+
+    scale:
+      replica: 2
+      auto: # automatic scale up and down based on given metrics
+        range: "1-10"
+        cpu: 80 # if cpu utilization is above 80%, scale up
+        qps: 1000 # if qps is higher than 1k, scale up
+
+    canary: # Auto-create canary deployment. Only upgrade after verify successfully.
+      replica: 1 # canary deployment size
+      headers:
+        - "foo:bar.*"
+```
+
+Save this file to project root dir, and run:
+
+```bash
+vela up
+```
+
+It will build container image, render deployment manifests in yaml, and apply them to the server.
+
+### Extensible Design
+
+The Appfile could be extended with more configurations by adding more capabilities to the OAM system. The config fields in Appfile are strongly correlated to the [capabilities system of OAM](https://github.com/oam-dev/kubevela/blob/master/DESIGN.md#capability-register-and-discovery) – Config fields are registered in the capabilities system and exposed via a [CUE template](https://cuelang.org/).
```

Comment:
Could you please also clarify how to express multiple components in single appfile?

Reply:
The proposal suggests a `services` field which implicitly implies multi components.
I think adding an example for it will be good enough.